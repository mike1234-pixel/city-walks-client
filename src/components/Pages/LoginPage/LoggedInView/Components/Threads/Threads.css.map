{"version":3,"sourceRoot":"","sources":["Threads.scss"],"names":[],"mappings":"AAAA;AAEA;EACE;EACA;EACA;EACA;EACA;EACA;;;AAGF;EACE;EACA;EACA;EACA;;;AAGF;EACE;;;AAGF;EACE;EACA","file":"Threads.css"}