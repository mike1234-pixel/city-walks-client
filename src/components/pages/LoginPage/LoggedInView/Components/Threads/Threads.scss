/** @format */

.threads-container {
  width: 100%;
  min-height: 70vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-bottom: 20px;
}

.add-thread-form-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.add-thread-form {
  width: 80vw;
}

.add-thread-form-heading {
  font-size: 20px;
  margin: 0 30px;
}
