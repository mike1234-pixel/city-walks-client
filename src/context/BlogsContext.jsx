import { createContext, useState } from "react"

export const BlogsContext = createContext();

export const BlogsContextProvider = (props) => {

    const [blogPosts, setBlogPosts] = useState([])
    const [blogsLoading, setBlogsLoading] = useState(true)

    return (
        <BlogsContext.Provider 
            value={{
                blogPosts,
                setBlogPosts,
                blogsLoading,
                setBlogsLoading,
            }}
        >
            {props.children}
        </BlogsContext.Provider>
    )
}