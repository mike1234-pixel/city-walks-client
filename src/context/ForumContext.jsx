import { createContext, useState } from "react"
import BoardBox from "../components/pages/LoginPage/LoggedInView/Components/BoardBox/BoardBox"

export const ForumContext = createContext();

export const ForumContextProvider = (props) => {
 
    const [loadingBoards, setLoadingBoards] = useState(true)
    const [boards, setBoards] = useState([])
    const [selectedThreads, setSelectedThreads] = useState([])

    const [currentBoardName, setCurrentBoardName] = useState("")
    const [currentBoardId, setCurrentBoardId] = useState("")

    let displayBoards;

    if (!loadingBoards) {
    displayBoards = boards.map((board, index) => {
        return (
            <BoardBox boardId={board._id} name={board.name} description={board.description} index={index} key={index} />
                )
        })
        // create routes for boards with Thread as the view component
    }

    return (
        <ForumContext.Provider 
            value={{
                setBoards,
                boards, // data
                setLoadingBoards,
                loadingBoards,
                displayBoards,
                setSelectedThreads,
                selectedThreads, 
                setCurrentBoardName,
                currentBoardName,
                setCurrentBoardId,
                currentBoardId,
            }}>
            {props.children}
        </ForumContext.Provider>
    )
}

