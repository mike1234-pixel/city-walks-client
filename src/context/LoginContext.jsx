import { createContext, useState } from "react"
import axios from "axios"
import qs from "qs"
import LoginForm from "../components/pages/LoginPage/LoginForm/LoginForm";
import RegistrationForm from "../components/pages/LoginPage/RegistrationForm/RegistrationForm"
import VerificationForm from "../components/pages/LoginPage/VerificationForm/VerificationForm";
import ResetPasswordForm from "../components/pages/LoginPage/ResetPasswordForm/ResetPasswordForm";
import ForgotPasswordForm from "../components/pages/LoginPage/ForgotPasswordForm/ForgotPasswordForm";
import Boards from "../components/pages/LoginPage/LoggedInView/Boards"

export const LoginContext = createContext();

export const LoginContextProvider = (props) => {
    const [popupVisible, setPopupVisible] = useState(false)

    const [firstName, setFirstName] = useState("")
    const [lastName, setLastName] = useState("")
    const [userId, setUserId] = useState("")
    const [registrationEmail, setRegistrationEmail] = useState("")
    const [registrationPassword, setRegistrationPassword] = useState("")
  
    const [loginEmail, setLoginEmail] = useState("")
    const [loginPassword, setLoginPassword] = useState("")
  
    // save all of this in local storage
    const [loggedIn, setLoggedIn] = useState(false)
    const [userFirstName, setUserFirstName] = useState("")
    const [userLastName, setUserLastName] = useState("")
  
    const [verificationEmail, setVerificationEmail] = useState("")

    const [resetPasswordEmail, setResetPasswordEmail] = useState("")
    const [resetPasswordOldPassword, setResetPasswordOldPassword] = useState("")
    const [resetPasswordNewPassword, setResetPasswordNewPassword] = useState("")
    const [resetPasswordConfirmNewPassword, setResetPasswordConfirmNewPassword] = useState("")

    const [forgotPasswordEmail, setForgotPasswordEmail] = useState("")

    const handlePopup = () => {
      setPopupVisible(false)
      localStorage.setItem("popupVisible", false)
    }

    // set display form on LoginPage
    const [form, setForm] = useState("loggedOutView");

    let displayForm = <Boards/>;
    if (form === "boards") {
      displayForm = <Boards/>;
    } else if (form === "loginForm") {
      displayForm = <LoginForm />;
    } else if (form === "registrationForm") {
      displayForm = <RegistrationForm />;
    } else if (form === "verificationForm") {
      displayForm = <VerificationForm />;
    } else if (form === "resetPasswordForm") {
      displayForm = <ResetPasswordForm />;
    } else if (form === "forgotPasswordForm") {
      displayForm = <ForgotPasswordForm />;
    }

    const handleChangeRegistration = (event) => {
        switch(event.target.name) {
            case "registration-fname":
              setFirstName(event.target.value)
              break;
            case "registration-lname":
              setLastName(event.target.value)
              break;
            case "registration-email":
              setRegistrationEmail(event.target.value)
              break;
            case "registration-password":
              setRegistrationPassword(event.target.value)
              break;
          } 
    }
  
    const handleSubmitRegistration = (event) => {
      console.log("handle submit triggered")
        event.preventDefault()
  
        const payload = {
            fname: firstName,
            lname: lastName,
            email: registrationEmail,
            password: registrationPassword,
          };
  
    axios
        .post("http://localhost:5000/register-user", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "An account with this email already exists.") {
            alert("An account with this email already exists.")
          } else if (res.data === "We have sent you an email. Please verify your account by clicking the link in the mail.") {
            alert("We have sent you an email. Please verify your account by clicking the link in the email. (This code expires after 10 minutes)")
           } else {
            alert("Registration Complete")
            setLoggedIn(true)
            setFirstName("")
            setLastName("")
            setRegistrationEmail("")
            setRegistrationPassword("")
            setUserFirstName(res.data.fname)
            setUserLastName(res.data.lname)
            window.scrollTo(0, 0)
          }
        });
    }
  
    const handleChangeLogin = (event) => {
      switch(event.target.name) {
          case "login-email":
            setLoginEmail(event.target.value)
            break;
          case "login-password":
            setLoginPassword(event.target.value)
            break;
        } 
  }
  
  const handleSubmitLogin = (event) => {
    console.log("handle submit login triggered")
      event.preventDefault()
  
      const payload = {
          email: loginEmail,
          password: loginPassword
        };
  
  axios
      .post("http://localhost:5000/login-user", qs.stringify(payload))
      .then((res, err) => {
        if (err) {
          console.log(err);
        } else if (res.data === "Your account exists but is not activated. Please click 'verify account' for email verification.") {
          alert("Your account exists but is not activated. Please click 'verify account' for email verification.")
        } else if (res.data === "unsuccessful login attempt") {
          alert("Unsuccessful Login Attempt. Please Try Again.")
        } else {
          alert("Login Complete")
          setLoggedIn(true)
          setLoginEmail("")
          setLoginPassword("")
          setForm("")
          setUserId(res.data._id)
          setUserFirstName(res.data.fname)
          setUserLastName(res.data.lname)
          localStorage.setItem("loggedIn", true)
          localStorage.setItem("userId", res.data._id)
          localStorage.setItem("userFirstName", res.data.fname)
          window.scrollTo(0, 0)
        }
      });
  }

  // resend verification email

  const handleChangeVerification = (event) => {
    switch(event.target.name) {
        case "verification-email":
          setVerificationEmail(event.target.value)
          break;
      }   
  }

const handleSubmitVerification = (event) => {
    console.log("handle submit verification")
    event.preventDefault()

    const payload = {
        email: verificationEmail,
      };

axios
    .post("http://localhost:5000/reverify-user", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("Verification email sent. Check your inbox.")
        setVerificationEmail("")
        window.scrollTo(0, 0)
      }
    })
}

  // reset password

  const handleChangeResetPassword = (event) => {
    switch(event.target.name) {
        case "reset-email":
          setResetPasswordEmail(event.target.value)
          break;
        case "old-password":
          setResetPasswordOldPassword(event.target.value)
          break;
        case "new-password":
          setResetPasswordNewPassword(event.target.value)
          break;
        case "confirm-new-password":
          setResetPasswordConfirmNewPassword(event.target.value)
          break;
      }   
  }

const handleSubmitResetPassword = (event) => {
    console.log("handle submit reset password")
    event.preventDefault()

    const payload = {
        email: resetPasswordEmail,
        oldPassword: resetPasswordOldPassword,
        newPassword: resetPasswordNewPassword,
      };

      if (resetPasswordNewPassword === resetPasswordConfirmNewPassword) {
        axios
        .post("http://localhost:5000/reset-password-with-old-password", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "user not found") {
            alert("We could not find your account. Please try again.")
          } else if (res.data === "old password does not match password in the database") {
            alert("The password you entered does not match the account. Please try again or use the Forgot Password option.")
          } else if (res.data === "password successfully updated") {
            alert("Password reset. You can now login.")
            setResetPasswordEmail("")
            setResetPasswordOldPassword("")
            setResetPasswordNewPassword("")
            setResetPasswordConfirmNewPassword("")
            window.scrollTo(0, 0)
          }
        })
      } else {
        alert("Passwords don't match. Please try again.")
      }

}

  // forgot password

  const handleChangeForgotPassword = (event) => {
    switch(event.target.name) {
        case "forgot-password-email":
          setForgotPasswordEmail(event.target.value)
          break;
      }   
  }

const handleSubmitForgotPassword = (event) => {
    console.log("handle submit forgot password")
    event.preventDefault()

    const payload = {
        email: forgotPasswordEmail,
      };

axios
    .post("http://localhost:5000/forgot-password", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("We have sent you an email. Please click the click in your email to reset your password")
        setForgotPasswordEmail("")
        window.scrollTo(0, 0)
      }
    })
}
  
  const logOut = () => {
    localStorage.clear()
    localStorage.setItem("popupVisible", false)
    setLoggedIn(false)
    setUserId("")
    setUserFirstName("")
    setUserLastName("")
    setForm("boards")
  }

    return (
        <LoginContext.Provider 
            value={{
              popupVisible,
              setPopupVisible,
              handlePopup,
               // display form
               displayForm,
               setForm,
               // logged in state 
               loggedIn,
               // registration state and functions
                firstName,
                lastName,
                registrationEmail,
                registrationPassword,
                handleChangeRegistration,
                handleSubmitRegistration,
                // login state and functions
                loginEmail,
                loginPassword,
                handleChangeLogin: handleChangeLogin,
                handleSubmitLogin: handleSubmitLogin,
                setLoggedIn,
                setUserFirstName,
                setUserId,
                // logged in user data
                userFirstName,
                userId,
                userLastName,
                // resend verification email
                verificationEmail,
                handleChangeVerification,
                handleSubmitVerification,
                // reset password
                resetPasswordEmail,
                resetPasswordOldPassword,
                resetPasswordNewPassword,
                resetPasswordConfirmNewPassword,
                handleChangeResetPassword,
                handleSubmitResetPassword,
                // forgot password
                forgotPasswordEmail,
                handleChangeForgotPassword,
                handleSubmitForgotPassword,
                // logout function
                logOut
            }}
        >
            {props.children}
        </LoginContext.Provider>
    )
}

// // setter
// localStorage.setItem('myData', data);
 
// // getter
// localStorage.getItem('myData');
 
// // remove
// localStorage.removeItem('myData');
 
// // remove all
// localStorage.clear();

