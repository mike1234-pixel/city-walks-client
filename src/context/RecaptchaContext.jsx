import { createContext, useEffect } from "react"

export const RecaptchaContext = createContext();

export const RecaptchaContextProvider = (props) => {
    
  const siteKey = "6LdmxiUaAAAAAIYySt3c8XvwOMokTQ_SW2cYkvMw";
 
  // dynamically load the javascript api using the site key
    useEffect(() => {
      const loadScriptByURL = (id, url, callback) => {
        const isScriptExist = document.getElementById(id);
    
        if (!isScriptExist) {
          var script = document.createElement("script");
          script.type = "text/javascript";
          script.src = url;
          script.id = id;
          script.onload = function () {
            if (callback) callback();
          };
          document.body.appendChild(script);
        }
    
        if (isScriptExist && callback) callback();
      }
    
      // load the script by passing the URL
      loadScriptByURL("recaptcha-key", `https://www.google.com/recaptcha/api.js?render=${siteKey}`, () => {
        console.log("recaptcha script loaded!");
      });
    }, []);

    return (
        <RecaptchaContext.Provider 
            value={{
            siteKey
            }}>
            {props.children}
        </RecaptchaContext.Provider>
    )
}