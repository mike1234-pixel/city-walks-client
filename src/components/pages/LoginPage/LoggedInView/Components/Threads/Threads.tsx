import React, { ReactNode, useEffect, useState } from "react"
import ThreadBox from "../ThreadBox/ThreadBox"
import toTitleCase from "../../../../../../functions/toTitleCase"
import { MDBBtn, MDBInput, MDBContainer } from "mdbreact"
import axios, { AxiosError, AxiosResponse } from "axios"
import qs from "qs"
import { connect } from 'react-redux'
import Board from '../../../../../../types/Boards/Board'
import GlobalState from "../../../../../../types/State/Global/State"
import "./Threads.scss"
import Thread from "../../../../../../types/PostRequests/Thread"
import ThreadT from '../../../../../../types/Boards/Thread'

interface Props {
    history: any;
    boards: Array<Board>;
    loggedIn: boolean;
    userId: string;
    userFirstName: string;
}

const Threads: React.FC<Props> = (props: Props) => {

    const { boards, loggedIn, userId, userFirstName } = props;

    const boardName: string = toTitleCase(props.history.location.pathname.replace("/forum/", "").replace(/-/g, " "))

    useEffect(() => {
        window.scrollTo(0, 0);
    }, []);

    let threads: string | ReactNode = "loading";

    if (boards.length) {

        const selectedBoard = boards.filter((board: Board) => board.name === boardName)[0]

        if (selectedBoard === undefined) {
            threads = "thread not found"
        } else {
            threads = selectedBoard.threads.map((thread: ThreadT, index: number) => {

                return (
                    <ThreadBox
                        currentBoardName={boardName}
                        threadId={thread._id}
                        userFirstName={thread.userFirstName}
                        title={thread.title}
                        content={thread.content}
                        replies={thread.replies}
                        submittedOn={thread.submittedOn}
                        userId={thread.userId}
                        key={index} loggedIn={false} currentUserFirstName={""} currentUserId={""} />
                )
            }).reverse()
        }
    }

    const [threadName, setThreadName] = useState<string>("")
    const [threadContent, setThreadContent] = useState<string>("")

    const handleChange: (event: React.ChangeEvent<any>) => void = (event) => {

        switch (event.target.name) {
            case "add-thread-name":
                setThreadName(event.target.value)
                break;
            case "add-thread-content":
                setThreadContent(event.target.value)
                break;
        }

    }

    const handleSubmit: (event: React.FormEvent) => void = (event) => {
        event.preventDefault()

        let payload: Thread = {
            currentBoardName: boardName,
            userId: userId,
            userFirstName: userFirstName,
            title: threadName,
            content: threadContent
        };

        axios
            .post("https://city-walks.herokuapp.com/add-thread", qs.stringify(payload))
            .then((res: AxiosResponse) => {

                alert("thread submitted.")
                setThreadName("")
                setThreadContent("")
                window.location.reload()

            }).catch((err: AxiosError) => {
                console.log(err)
            });
    }

    const addThread: ReactNode =
        <div className="add-thread-form-container">
            <h2 className="add-thread-form-heading">Post a new discussion thread on this board:</h2>
            <form className="add-thread-form" onSubmit={handleSubmit}>
                <MDBInput type="text" name="add-thread-name" id="add-thread-name" onChange={handleChange} value={threadName} label="thread name" />
                <MDBInput type="textarea" rows="6" name="add-thread-content" id="add-thread-content" onChange={handleChange} value={threadContent} label="thread content" />
                <MDBBtn outline color="elegant" type="submit">Add Thread</MDBBtn>
            </form>
        </div>

    return (
        <MDBContainer>
            <div className="threads-container">
                <h1 className="page-heading">{boardName}</h1>
                {loggedIn && addThread}
                <div>{threads}</div>
            </div>
        </MDBContainer>
    )
}

const mapStateToProps = (state: GlobalState) => ({
    boards: state.boardsState.boards,
    loggedIn: state.loginState.loggedIn,
    userId: state.loginState.userId,
    userFirstName: state.loginState.userFirstName
});

export default connect(mapStateToProps, null)(Threads);