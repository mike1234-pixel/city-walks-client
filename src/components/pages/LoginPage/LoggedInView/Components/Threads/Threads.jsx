import { useContext, useEffect, useState } from "react"
import { ForumContext } from "../../../../../../context/ForumContext"
import { LoginContext } from "../../../../../../context/LoginContext"
import ThreadBox from "../ThreadBox/ThreadBox"
import toTitleCase from "../../../../../../functions/toTitleCase"
import { MDBBtn, MDBInput } from "mdbreact"
import axios from "axios"
import qs from "qs"
import PopUp from "../../../../../PopUp/PopUp"
import "./Threads.css"

const Threads = ({match}) => {

    const { boards, loadingBoards } = useContext(ForumContext)
    const { loggedIn, userId, userFirstName, popupVisible } = useContext(LoginContext)

    const boardName = toTitleCase(match.url.replace("/boards/", "").replace(/-/g, " "))

    useEffect(() => {
        window.scrollTo(0, 0);
        }, []);

    let threads = "loading";
    
    if (!loadingBoards) {

    const selectedBoard = boards.filter((board) => board.name === boardName)[0]

    if (selectedBoard === undefined) {
        threads = "thread not found"
    } else {
        
        threads = selectedBoard.threads.map((thread, index) => {
            return (
            <ThreadBox 
                currentBoardName={boardName}
                threadId={thread._id}
                userFirstName={thread.userFirstName}
                title={thread.title}
                content={thread.content}
                replies={thread.replies}
                submittedOn={thread.submittedOn}
                userId={thread.userId}
                key={index}
                />
            )
        }).reverse()
    }
}

    const [threadName, setThreadName] = useState("")
    const [threadContent, setThreadContent] = useState("")

    const handleChange = (event) => {

        switch(event.target.name) {
            case "add-thread-name":
              setThreadName(event.target.value)
              break;
            case "add-thread-content":
              setThreadContent(event.target.value)
              break;
        }
   
    }

    const handleSubmit = (event) => {
        event.preventDefault()

        let payload = {
            currentBoardName: boardName,
            userId: userId,
            userFirstName: userFirstName,
            title: threadName,
            content: threadContent
          };

        axios
        .post("http://localhost:5000/add-thread", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else {
              alert("thread submitted.")
              setThreadName("")
              setThreadContent("")
              window.location.reload()
          }
        });
    }

const addThread = 
<div className="add-thread-form-container">
<h2 className="add-thread-form-heading">Post a new discussion thread on this board:</h2>
<form className="add-thread-form" onSubmit={handleSubmit}>
    <MDBInput type="text" name="add-thread-name" id="add-thread-name" onChange={handleChange} value={threadName} label="thread name"/>
    <MDBInput type="textarea" rows="6" name="add-thread-content" id="add-thread-content" onChange={handleChange} value={threadContent} label="thread content"/>
    <MDBBtn outline color="elegant" type="submit">Add Thread</MDBBtn>
</form>
</div>

    return (
        <div className="threads-container">
            {popupVisible && <PopUp/>}
            <h1 className="page-heading">{boardName}</h1>
                {loggedIn && addThread}
            <div>{threads}</div>
        </div>
    )
}

export default Threads