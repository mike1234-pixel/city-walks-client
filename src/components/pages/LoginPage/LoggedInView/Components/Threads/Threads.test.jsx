import renderer from 'react-test-renderer'
import { LoginContextProvider } from "../../../../../../context/LoginContext"
import { ForumContextProvider } from "../../../../../../context/ForumContext"
import Threads from "./Threads"

// snapshot test
it('Threads renders correctly', () => {  

  const tree = renderer
    .create(
        <ForumContextProvider>
            <LoginContextProvider>
                <Threads match={{params: {id: 1}, isExact: true, path: "", url: "/boards/your-city"}} />
            </LoginContextProvider>
        </ForumContextProvider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});