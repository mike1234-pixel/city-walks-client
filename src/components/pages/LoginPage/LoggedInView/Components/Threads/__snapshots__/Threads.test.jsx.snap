// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Threads renders correctly 1`] = `
<div
  className="threads-container"
>
  <h1
    className="page-heading"
  >
    Your City
  </h1>
  <div>
    loading
  </div>
</div>
`;
